xds/xds.go

```diff
@@ -74,22 +77,69 @@ func init() {
 		v3statusgrpc.RegisterClientStatusDiscoveryServiceServer(grpcServer, csdss)
 		return csdss.Close, nil
 	})
+
+	internal.NewXDSResolverWithConfigForTesting = func(bootstrapConfig []byte) (resolver.Builder, error) {
+		return xdsresolver.NewBuilderForTesting(bootstrapConfig)
+	}
+}
+
+// RegisterRLSClusterSpecifierPluginForTesting registers the RLS Cluster
+// Specifier Plugin for testing purposes, regardless of the XDSRLS environment
+// variable.
+//
+// TODO: Remove this function once the RLS env var is removed.
+//
+// Testing Only
+//
+// This function should ONLY be used for testing purposes.
+//
+// Experimental
+//
+// Notice: This API is EXPERIMENTAL and will be removed in a later release.
+func RegisterRLSClusterSpecifierPluginForTesting() {
+	rls.RegisterForTesting()
 }
 
-// NewXDSResolverWithConfigForTesting creates a new xds resolver builder using
-// the provided xds bootstrap config instead of the global configuration from
-// the supported environment variables.  The resolver.Builder is meant to be
-// used in conjunction with the grpc.WithResolvers DialOption.
+// UnregisterRLSClusterSpecifierPluginForTesting unregisters the RLS Cluster
+// Specifier Plugin for testing purposes. This is needed because there is no way
+// to unregister the RLS Cluster Specifier Plugin after registering it solely
+// for testing purposes using RegisterRLSClusterSpecifierPluginForTesting().
+//
+// TODO: Remove this function once the RLS env var is removed.
 //
 // Testing Only
 //
-// This function should ONLY be used for testing and may not work with some
-// other features, including the CSDS service.
+// This function should ONLY be used for testing purposes.
 //
 // Experimental
 //
-// Notice: This API is EXPERIMENTAL and may be changed or removed in a
-// later release.
-func NewXDSResolverWithConfigForTesting(bootstrapConfig []byte) (resolver.Builder, error) {
```

Comment:
Why was this moved to `internal` but these other, temporary, RLS/RBAC testing functions were added here?

Reply:
This function is used from a couple of places: `internal/testutils/xds/e2e` and `test/xds` while the other temporary register/unregister funcs are only used from `test/xds`. There are still some tests in `xds/internal/` which use the `internal/testutils/xds/e2e` package. So, there were some import loops. For example: top-level xds package --> xds/csds --> internal/testutils/xds/e2e --> top-level-xds-package (for the NewXDSResolverWithConfigForTesting func).

I will try to move those tests out after this PR.

Reply:
Couldn't these also live in `internal`, though, just like the new `NewXDSResolverWithConfigForTesting`?  That should avoid import loops, too, right?

Reply:
The import loop was only with `NewXDSResolverWithConfigForTesting`, which is why I moved it to the `internal` package. I felt like moving the other ones to the `internal` package would add too many symbols to it. Please let me know if you feel strongly about this. Thanks.

Reply:
IMO it's better for them to be in `internal` than visible in our external godoc.  I should have originally done `NewXDSResolver...` that way in hindsight.  I imagine it shouldn't be that hard to move these, but if it is, then you can leave them because they're also expected to be temporary.

Also, would this work??

```go
package internal

var NewXDSResolverBlah interface{}  // as always

.....

package internal/xds

func NewXDSResolverBlah(args) returns {
	// optionally handle `nil`??  Or can we actually import the thing that sets it from here??  Seems plausible.
	return internal.NewXDSResolverBlah(func(args)returns)(args)
}
```

That way we don't have to type cast the functions whenever we use them.  This could be a follow-up cleanup, and we could even do something similar with all the other stuff in `internal` that needs to be type cast every time we use it.

Reply:
> Or can we actually import the thing that sets it from here??  

I guess in the xds cases, "no", because of the visibility problem.  But in the other cases (e.g. ParseServiceConfig) then I think we could.

Reply:
Moved everything to `internal` package. Thanks.